pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/impl/AbstractNode.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ast.impl;
+
+import org.apache.commons.lang3.ArrayUtils;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import net.sourceforge.pmd.lang.ast.Node;
+import net.sourceforge.pmd.util.DataMap;
+import net.sourceforge.pmd.util.DataMap.DataKey;
+
+/**
+ * Base class for implementations of the Node interface whose children
+ * are stored in an array. This class provides the basic utilities to
+ * link children and parent. It's used by most most nodes, but currently
+ * not the antlr nodes, so downcasting {@link Node} to this class may fail
+ * and is very bad practice.
+ *
+ * @param <B> Self type (eg AbstractJavaNode in the java module), this
+ *            must ultimately implement {@code <N>}, though the java type
+ *            system does not allow us to express that
+ * @param <N> Public interface for nodes of this language (eg JavaNode
+ *            in the java module).
+ */
+public abstract class AbstractNode<B extends AbstractNode<B, N>, N extends GenericNode<N>> implements GenericNode<N> {
+
+    private static final Node[] EMPTY_ARRAY = new Node[0];
+
+    // lazy initialized, many nodes don't need it
+    private @Nullable DataMap<DataKey<?, ?>> userData;
+
+    // never null, never contains null elements
+    private Node[] children = EMPTY_ARRAY;
+    private B parent;
+    private int childIndex;
+
+    protected AbstractNode() {
+        // only for subclassing
+    }
+
+    @Override
+    public N getParent() {
+        return (N) parent;
+    }
+
+    @Override
+    public int getIndexInParent() {
+        return childIndex;
+    }
+
+    @Override
+    public N getChild(final int index) {
+        return (N) children[index];
+    }
+
+    @Override
+    public int getNumChildren() {
+        return children.length;
+    }
+
+    protected void setParent(final B parent) {
+        this.parent = parent;
+    }
+
+    @SuppressWarnings("unchecked")
+    private B asSelf(Node n) {
+        return (B) n;
+    }
+
+    /**
+     * Set the child at the given index to the given node. This resizes
+     * the children array to be able to contain the given index. Implementations
+     * must take care that this does not leave any "holes" in the array.
+     * This method throws if there is already a child at the given index.
+     *
+     * <p>Note that it is more efficient to add children in reverse
+     * (from right to left), because the array is resized only the
+     * first time.
+     *
+     * <p>This method also calls {@link #setParent(AbstractNode)}.
+     *
+     * @param child The child to add
+     * @param index The index to which the child will be added
+     */
+    protected void addChild(final B child, final int index) {
```

Comment:
setChildren in used only in tests AFAIK so could be pulled down to DummyNode. 

I think addChild is a bit unsafe/hard to use:
* nothing prevents you from writing `addChild(child, 42)`, and leave 42 `null` slots in the array (which breaks the contract of the class)... 
* adding children with this method needs to be done right to left, otherwise adding `n` children performs `n` array resizes

OTOH I can't see another way to improve it right now apart from documentation